### DISCUSSION TITLE
Find a way to use CubeMX generated code inside Arduino IDE or PlatformIO.

### DISCUSSION THREAD

**Discussion by HamidSaffari** [stm32duino/Arduino_Core_STM32#1373] [Ideas]
Hi and best regards.
Since there is many variants of ST MCUs and take times to add all of them on Arduino IDE,
and also can optimize code and so,
I wonder if we can find a way to use CubeMX ( STM32CubeMX ) generated code inside Arduino IDE or PlatformIO and link them together.

**Comment by fpistm** [comment]
Hi @HamidSaffari 
I've converted this topic to GH discussion as it is not an issue.

I agree, that would be fine. As only the generic clock config and linker script are missing.

About the clock config, files provided by [STM32_open_pin_data](https://github.com/STMicroelectronics/STM32_open_pin_data/) provide some information 
https://github.com/STMicroelectronics/STM32_open_pin_data/blob/d1b80258bc63add9baad70be0ef74f3c4301ac58/mcu/STM32F411C(C-E)Ux.xml#L3

For example here :
`Mcu ClockTree="STM32F411"`

The GH does not provide all the database anyway in CubeMx one in `db/mcu/IP` we could find `RCC-STM32F411-rcc_v1_0_Modes.xml` which provide several clock informations but uses it seems not so easy.
Anyway processing this is not easy. A best way would be to be able to use CubeMX in CLI mode to access the clock solver and also linker script generation.

**Comment by HamidSaffari** [comment]
Thanks.
I find How to customize system clock by STM32CubeMX and replace the variant.cpp `SystemClock_Config(void)` func by src/main.c same func [here](https://github.com/stm32duino/wiki/wiki/Add-a-new-variant-(board)#5---system-clock-configuration)

And also the way to make it works on platformIO from [here](https://community.platformio.org/t/using-stm32cubemx-and-platformio/2611/2) based on Krishna_Chaitanya which make it so simple and meant-to-be but for stm32cube framework and not Arduino.

Based on Krishna_Chaitanya:
Create a new project in stm32cubeMX app:

![daa62a999972162196a65f8fa680139d145f673f](https://user-images.githubusercontent.com/35362339/116430490-223fdb80-a85c-11eb-8ad7-6082f2833969.png)
I am choosing the STM32F030R8 for my demo and click on “start project”

Under Project > Settings

![969751e49be028e28186b81846fab6b21348c41b](https://user-images.githubusercontent.com/35362339/116430590-3aaff600-a85c-11eb-8c8a-8a4dc687ec61.png)
Select Toolchain / IDE as “Other toolchain (GPDSC)” under the Project tab

![5447bbb990053fb74cd576ce573c4a59805665ea](https://user-images.githubusercontent.com/35362339/116430793-67fca400-a85c-11eb-82f2-5b0b50bc6f32.png)
Under Code Generator tab, select the “Copy only the necessary library files” and “Generate peripheral initialization as a pair of .c / .h files per peripheral”

![3e7270bad192d7b511aa2e093dc8a17acdefa18e](https://user-images.githubusercontent.com/35362339/116430928-85317280-a85c-11eb-8d3e-20b6784b12f6.png)
Apply the settings, customize the pinouts, peripherals as per your need. When you are done initializing the MCU, click on the “Generate Code” option.
This is how the project output looks like. A bunch of headers and C files under the “Inc” and “Src” folders.

![e5c3dc0be2e01e2e964a0f90ad472ea2c55a3654](https://user-images.githubusercontent.com/35362339/116431011-98dcd900-a85c-11eb-9901-f3be8e870123.png)
Now copy the headers and source files under the “src” folder in the Platformio IDE and compile.
Thats all! : It should compile without any errors or issues.
(Well, at least it did so for me)

```
 [env:disco_f030r8]
 platform = ststm32
board = disco_f030r8
framework = stm32cube
build_flags = -ICore/Inc
```
Now it works for stm32cube framework but not Arduino framework.
